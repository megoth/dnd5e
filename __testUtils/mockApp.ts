import { AppModel, createApp, getBundleKey } from "../src/models/app";
import mockFluentBundle from "./mockFluentBundle";
import { appVocabURL } from "./mockAppIndexDataset";
import mockLanguage, { defaultLocale } from "./mockLanguage";
import mockResourceBundleMap from "./mockResourceBundleMap";
import mockAppIndex from "./mockAppIndex";

export const errorsURL = "https://example.com/global/errors.ttl";
export const faqsURL = "https://example.com/global/faqs.ttl";
export const translationsURL = "https://example.com/global/translations.ttl";
export const localizationsURL =
  "https://example.com/global/translations.en-US.ttl";

export function mockAppPropValue(val, bundleName = "global", locale = "en-US") {
  return bundleName
    ? {
        [getBundleKey(locale, bundleName)]: val,
      }
    : {
        [locale]: val,
      };
}

export default function mockApp(overrides: Partial<AppModel> = {}): AppModel {
  const currentLocale = overrides.currentLocale || defaultLocale;
  return {
    appVocabURL: overrides.appVocabURL || appVocabURL,
    currentLocale,
    fluentBundles: {
      [currentLocale]: mockFluentBundle({}, translationsURL, {
        locale: currentLocale,
      }),
      ...overrides.fluentBundles,
    },
    languages: [mockLanguage(currentLocale), ...(overrides.languages || [])],
    resourceBundles: {
      ...mockResourceBundleMap({
        label: "admin",
        locale: currentLocale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
      ...mockResourceBundleMap({
        locale: currentLocale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
      ...overrides.resourceBundles,
    },
  };
}

export function mockNorwegianApp(overrides: Partial<AppModel> = {}): AppModel {
  const locale = "nb-NO";
  return mockApp({
    currentLocale: locale,
    fluentBundles: {
      [locale]: mockFluentBundle({}, translationsURL, {
        locale,
      }),
    },
    languages: [
      mockLanguage(locale, {
        languageCode: "ðŸ‡³ðŸ‡´",
      }),
    ],
    resourceBundles: {
      ...mockResourceBundleMap({
        locale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
    },
    ...overrides,
  });
}

export function mockEmptyApp(): AppModel {
  return createApp(defaultLocale, appVocabURL, null);
}

export function mockUnloadedApp(): AppModel {
  return createApp(defaultLocale, appVocabURL, mockAppIndex());
}
