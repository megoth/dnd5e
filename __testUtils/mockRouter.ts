import { NextRouter } from "next/router";

export default function mockRouter(
  overrides: Partial<NextRouter> = {}
): NextRouter {
  return {
    asPath: "/",
    query: {},
    ...overrides,
  } as NextRouter;
}
