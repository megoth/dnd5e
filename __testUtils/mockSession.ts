import { Session } from "@inrupt/solid-client-authn-browser";

export interface SessionInfo {
  session: Session;
  sessionRequestInProgress: boolean;
  fetch: typeof window.fetch;
  login: () => Promise<void>;
  logout: () => Promise<void>;
}

export const authenticatedWebId = "http://example.com/webId#me";

export function mockAuthenticatedSession(options: Partial<Session> = {}) {
  return ({
    session: {
      info: {
        isLoggedIn: true,
        sessionId: "some-authenticated-session-id",
        webId: authenticatedWebId,
      },
      sessionRequestInProgress: false,
    },
    handleIncomingRedirect: () => Promise.resolve(),
    login: () => Promise.resolve(),
    logout: () => Promise.resolve(),
    on: () => {},
    ...options,
  } as unknown) as SessionInfo;
}

export function mockUnauthenticatedSession(options: Partial<Session> = {}) {
  return ({
    session: {
      info: {
        isLoggedIn: false,
        sessionId: "some-unauthenticated-session-id",
        webId: null,
      },
      sessionRequestInProgress: false,
    },
    handleIncomingRedirect: () => Promise.resolve(),
    login: options.login || jest.fn().mockResolvedValue(undefined),
    logout: () => Promise.resolve(),
    on: () => {},
    ...options,
  } as unknown) as SessionInfo;
}
