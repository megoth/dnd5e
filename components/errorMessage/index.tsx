import React from "react";
import { Localized } from "@fluent/react";
import NestedError from "nested-error-stacks";
import { getErrorId } from "../../src/models/error";

interface Props {
  error: NestedError;
}

export const TESTID_ERROR = "error";
export const TESTID_ERROR_TITLE = "error-title";

export default function ErrorMessage({ error }: Props) {
  const { message: errorUrl } = error;
  const errorId = getErrorId(errorUrl);
  return (
    <div data-testid={TESTID_ERROR}>
      <h1 data-testid={TESTID_ERROR_TITLE}>
        <span resource={errorUrl}>
          <Localized id={errorId}>{errorUrl}</Localized>
        </span>
      </h1>
    </div>
  );
}
