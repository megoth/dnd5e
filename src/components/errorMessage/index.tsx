import React from "react";
import NestedError from "nested-error-stacks";

interface Props {
  error: NestedError;
}

export const TESTID_ERROR = "error";
export const TESTID_ERROR_TITLE = "error-title";

export default function ErrorMessage({ error }: Props) {
  // const { message: errorURL } = error;
  return (
    <div data-testid={TESTID_ERROR}>
      <h1 data-testid={TESTID_ERROR_TITLE}>
        {/*{translationURL ? (*/}
        {/*  <Localized id={getTranslationId(translationURL)}>*/}
        {/*    {errorURL}*/}
        {/*  </Localized>*/}
        {/*) : (*/}
        {error.message}
        {/*)}*/}
      </h1>
    </div>
  );
}
