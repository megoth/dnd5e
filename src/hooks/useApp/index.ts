import { useContext, useEffect } from "react";
import { AppContext } from "../../contexts/app";
import { appIsLoadingLocalizations } from "../../models/app";

export default function useApp(bundles: string[] = []) {
  const { app, setBundles } = useContext(AppContext);
  useEffect(() => {
    if (!app || !appIsLoadingLocalizations(app, bundles)) {
      return;
    }
    setBundles(bundles);
  }, [app, bundles, setBundles]);
  return app;
}
