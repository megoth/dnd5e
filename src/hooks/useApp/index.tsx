import { useContext } from "react";
import AppContext from "./context";

export default function useApp() {
  return useContext(AppContext);
}
