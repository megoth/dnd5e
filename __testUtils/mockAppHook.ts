import mockApp from "./mockApp";

export default function mockAppHook(mock: jest.Mock, app = mockApp()) {
  mock.mockReturnValue(app);
}
