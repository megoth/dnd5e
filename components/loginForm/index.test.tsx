import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import * as solidUIReactFns from "@inrupt/solid-ui-react";
import * as routerFns from "next/router";
import useApp from "../../src/hooks/useApp";
import mockAppHook from "../../__testUtils/mockAppHook";
import { appName } from "../../__testUtils/mockApp";
import { mockUnauthenticatedSession } from "../../__testUtils/mockSession";
import { TESTID_ERROR } from "../errorMessage";
import LoginForm, {
  TESTID_LOGIN_FORM_BUTTON,
  TESTID_LOGIN_FORM_IDP_FIELD,
  TESTID_LOGIN_FORM_REMEMBER_CHECKBOX,
} from "./index";
import mockRouter from "../../__testUtils/mockRouter";

jest.mock("../../src/hooks/useApp");
const mockedAppHook = useApp as jest.Mock;

describe("LoginForm", () => {
  const idpUrl = "https://example.com";
  const idp1 = "http://example1.com";
  const idp2 = "http://example2.com";

  let unauthenticatedSession;
  let mockedSessionHook;
  let mockedRouterHook;

  beforeEach(() => mockAppHook(mockedAppHook));
  beforeEach(() => {
    unauthenticatedSession = mockUnauthenticatedSession();
    mockedSessionHook = jest
      .spyOn(solidUIReactFns, "useSession")
      .mockReturnValue(unauthenticatedSession);
  });
  beforeEach(() => {
    mockedRouterHook = jest
      .spyOn(routerFns, "useRouter")
      .mockReturnValue(mockRouter());
  });

  it("renders", () => {
    const { asFragment } = render(<LoginForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("offers a form to initiate authentication process", () => {
    const { getByTestId } = render(<LoginForm />);
    const idpField = getByTestId(TESTID_LOGIN_FORM_IDP_FIELD);
    userEvent.type(idpField, `${idpUrl}{enter}`);
    const { login } = unauthenticatedSession;
    expect(login).toHaveBeenCalledWith({
      clientName: appName,
      oidcIssuer: idpUrl,
      redirectUrl: "http://localhost/",
    });
  });

  it("offers the option to remember the IdP for later", () => {
    const { getByTestId } = render(<LoginForm />);
    const rememberCheckbox = getByTestId(TESTID_LOGIN_FORM_REMEMBER_CHECKBOX);
    userEvent.click(rememberCheckbox);
    const idpField = getByTestId(TESTID_LOGIN_FORM_IDP_FIELD);
    userEvent.type(idpField, `${idpUrl}{enter}`);
    expect(localStorage.getItem("rememberIdP")).toEqual("true");
    expect(localStorage.getItem("idp")).toEqual(idpUrl);
  });

  it("restores states from localStorage if opted-in", () => {
    localStorage.setItem("rememberIdP", "true");
    localStorage.setItem("idp", idpUrl);
    const { getByTestId } = render(<LoginForm />);
    const rememberCheckbox = getByTestId(TESTID_LOGIN_FORM_REMEMBER_CHECKBOX);
    expect(rememberCheckbox.getAttribute("checked")).toBeDefined();
    const idpField = getByTestId(TESTID_LOGIN_FORM_IDP_FIELD);
    expect(idpField.getAttribute("value")).toEqual(idpUrl);
  });

  it("does not remember idp if remember me is not checked", () => {
    localStorage.setItem("rememberIdP", "true");
    localStorage.setItem("idp", idpUrl);
    const { getByTestId } = render(<LoginForm />);
    const rememberCheckbox = getByTestId(TESTID_LOGIN_FORM_REMEMBER_CHECKBOX);
    userEvent.click(rememberCheckbox);
    const submitButton = getByTestId(TESTID_LOGIN_FORM_BUTTON);
    userEvent.click(submitButton);
    expect(localStorage.getItem("rememberIdP")).toEqual("false");
    expect(localStorage.getItem("idp")).toBeNull();
  });

  it("renders an error if something goes wrong", () => {
    mockedSessionHook.mockReturnValue(
      mockUnauthenticatedSession({
        login: () => {
          throw new Error();
        },
      })
    );
    const { getByTestId } = render(<LoginForm />);
    const idpField = getByTestId(TESTID_LOGIN_FORM_IDP_FIELD);
    userEvent.type(idpField, `${idpUrl}{enter}`);
    expect(getByTestId(TESTID_ERROR)).toBeDefined();
  });

  it("can provide IdP via the query param", () => {
    mockedRouterHook.mockReturnValue({
      asPath: "/rules",
      query: { idp: idp1 },
    });
    const { getByTestId } = render(<LoginForm />);
    expect(
      getByTestId(TESTID_LOGIN_FORM_IDP_FIELD).getAttribute("value")
    ).toEqual(idp1);
  });

  it("also handles an array of IdPs", () => {
    mockedRouterHook.mockReturnValue({
      asPath: "/rules",
      query: { idp: [idp1, idp2] },
    });
    const { getByTestId } = render(<LoginForm />);
    expect(
      getByTestId(TESTID_LOGIN_FORM_IDP_FIELD).getAttribute("value")
    ).toEqual(idp1);
  });

  it("focuses on idp field if IdP query is given", () => {
    mockedRouterHook.mockReturnValue({
      asPath: "/rules",
      query: { idp: [idp1, idp2] },
    });
    const { getByTestId } = render(<LoginForm />);
    expect(getByTestId(TESTID_LOGIN_FORM_IDP_FIELD)).toEqual(
      document.activeElement
    );
  });
});
