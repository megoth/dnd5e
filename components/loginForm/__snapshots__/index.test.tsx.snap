// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm renders 1`] = `
<DocumentFragment>
  <form
    class="mt-2 md:mt-0 text-left"
  >
    <img
      alt="[Translation for https://example.com/global/translations.ttl#loginImage does not exist]"
      class="mx-auto hidden md:block"
      src="/logos/solid-emblem.svg"
      style="max-width: 200px;"
    />
    <label
      class="label"
      for="idp"
    >
      [Translation for https://example.com/global/translations.ttl#solidIdP does not exist]
      <input
        class="input"
        data-testid="login-form-idp-field"
        id="idp"
        placeholder="https://"
        required=""
        type="url"
        value=""
      />
    </label>
    <label
      class="label"
      for="rememberIdp"
    >
      <input
        class="checkbox"
        data-testid="login-form-remember-checkbox"
        id="rememberIdp"
        type="checkbox"
      />
      <span
        class="label__text"
      >
        [Translation for https://example.com/global/translations.ttl#rememberIdP does not exist]
      </span>
    </label>
    <button
      class="button w-full"
      data-testid="login-form-button"
      type="submit"
    >
      [Translation for https://example.com/global/translations.ttl#login does not exist]
    </button>
    <div
      class="mt-4"
    >
      <a
        class="link"
        href="/signup"
      >
        [Translation for https://example.com/global/translations.ttl#signupSolidPrompt does not exist]
      </a>
    </div>
  </form>
</DocumentFragment>
`;
